content/4_Gold/TopoSort.mdx

```diff
@@ -190,193 +214,162 @@ MAX_N = 10**5
 sys.setrecursionlimit(MAX_N)
 
 
-def dfs(node):
-	for i in graph[node]:
-		if not visited[i]:
-			visited[i] = True
-			dfs(i)
+def dfs(node: int) -> None:
+	for n in graph[node]:
+		if not visited[n]:
+			visited[n] = True
+			dfs(n)
 	top_sort.append(node)
 
 
-# N is number of nodes
-# M is number of edges
-N, M = map(int, input().strip().split())
-graph = [[] for _ in range(N)]
-top_sort = []
-ind = [0] * N
+# The number of nodes and edges respectively
+n, m = map(int, input().strip().split())
+graph = [[] for _ in range(n)]
 
-for i in range(M):
-	u, v = map(int, input().strip().split())
-	graph[u - 1].append(v - 1)
+for _ in range(m):
+	a, b = map(int, input().strip().split())
+	graph[a - 1].append(b - 1)
 
-visited = [False] * N
-for i in range(N):
+visited = [False] * n
+top_sort = []
+for i in range(n):
 	if not visited[i]:
 		visited[i] = True
 		dfs(i)
-
 top_sort = top_sort[::-1]
-for i in range(N):
+
+ind = [0] * n
+for i in range(n):
 	ind[top_sort[i]] = i
 
 # Check if the topological sort is valid
-flag = 0
-for i in range(N):
+valid = True
+for i in range(n):
 	for j in graph[i]:
 		if ind[j] <= ind[i]:
-			flag = 1
-			print("IMPOSSIBLE")
+			valid = False
 			break
 
-	if flag == 1:
+	if not valid:
 		break
 
-top_sort = [x + 1 for x in top_sort]
-if flag == 0:
-	print(*top_sort)
+if valid:
+	print(*[i + 1 for i in top_sort])
+else:
+	print("IMPOSSIBLE")
 ```
 
 </PySection>
-
 </LanguageSection>
 
-### Finding a Cycle
-
-<FocusProblem problem="dir" />
 
-We can modify the algorithm above to return a directed cycle in the case where a
-topological sort does not exist. To find the cycle, we add each node we visit
-onto the stack until we detect a node already on the stack.
+## BFS
 
-For example, suppose that our stack currently consists of $s_1,s_2,\ldots,s_i$
-and we then visit $u=s_j$ for some $j\le i$. Then
-$s_j\to s_{j+1}\to \cdots\to s_i\to s_j$ is a cycle. We can reconstruct the
-cycle without explicitly storing the stack by marking $u$ as not part of the
-stack and recursively backtracking until $u$ is reached again.
+The BFS version is known as
+[Kahn's Algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm).
```

Comment:
maybe a brief description on how it works? You and I both know wikipedia isnt pleasant to read

Reply:
this pr is mostly for the code